A17-client-side-health-checking.md

```diff
@@ -0,0 +1,317 @@
+via centralized health checking via a look-aside load balancer, where
+the health checks come from a small number of clients and where there
+is existing infrastructure to periodically poll each client.  However,
+if we are going to have large numbers of clients issuing health-check
+requests, then we will need to convert the health-checking protocol to
+a streaming Watch-based API for scalability and bandwidth-usage reasons.
+
+Note that one down-side of this approach is that it could conceivably
+be possible that the server-side health-checking code somehow fails
+to send an update when becoming unhealthy.  If the problem is due to
+a server that stops polling for I/O, then the problem would be caught
+by keepalives, at which point the client would disconnect.  But if the
+problem is caused by a bug in the health-checking service, then it's
+possible that a server could still be responding but has failed to notify
+the client that it is unhealthy.
+
+#### API Changes
+
+For the proposed API, see https://github.com/grpc/grpc-proto/pull/33.
+
+We propose to add the following RPC method to the health checking service:
+
+```
+rpc Watch(HealthCheckRequest) returns (stream HealthCheckResponse);
+```
+
+This is exactly the same as the existing `Check()` method, except that it
+is server-side streaming instead of unary.  The server will be expected
+to send a message immediately upon receiving the message from the client
+and will then be expected to send a new message whenever the requested
+service's health changes.
+
+We will also add a new enum value to the `ServingStatus` enum:
+
+```
+SERVICE_UNKNOWN = 3;
+```
+
+This serving status will be returned by the `Watch()` method when the
+requested service name is initially unknown by the server.  Note that the
+existing `Check()` unary method fails with status `NOT_FOUND` in this case,
+and we are not proposing to change that behavior.
+
+### Client Behavior
+
+Client-side health checking will be disabled by default;
+service owners will need to explicitly enable it via the [service
+config](#service-config-changes) when desired.  There will be a channel
+argument that can be used on the client to disable health checking even
+if it is enabled in the service config.
+
+The health checking client code will be built into the subchannel,
+so that individual LB policies do not need to explicitly add support
+for health checking. (However, see also "LB Policies Can Disable Health
+Checking When Needed" below.)
+
+Note that when an LB policy sees a subchannel go into state
+`TRANSIENT_FAILURE`, it will not be able to tell whether the subchannel
+has become disconnected or whether the backend has reported that it
+is unhealthy.  The LB policy may therefore take some unnecessary but
+unharmful steps in response, such as asking the subchannel to connect
+(which will be a no-op if it's already connected) or asking the resolver
+to re-resolve (which will probably be throttled by the cooldown timer).
```

Comment:
A possible solution here is to add another state called "connected" between "connecting" and "ready". The "connected" state tells LB policy to not re-resolve or ask the sub-channel to reconnect. Of course, this will increase the complexity of the state machine and also a re-resolution may anyway be necessary.

Reply:
These states are part of our public API in all languages; they are directly visible to applications.  So to add a new state, we would need to either (a) add the new state to the public API, thus unnecessarily exposing internal implementation details to the application, or (b) use a different set of states when internally propagating the state than when exposing it to the application, which also adds complexity to our code.  I think either of those options would be a fair amount of trouble, and that doesn't seem worthwhile, given that the extra work being done here is unharmful.

If this extra work does turn out to be harmful at some point in the future, I think we can find ways to avoid it without having to add a new connectivity state.